pkg/network/tracer/connection/ebpfless/tcp_processor.go

```diff
@@ -0,0 +1,190 @@
+	hasRemoteFin bool
+	// localFinSeq is the tcp.Seq number for the outgoing FIN (including any payload length)
+	localFinSeq uint32
+	// remoteFinSeq is the tcp.Seq number for the incoming FIN (including any payload length)
+	remoteFinSeq uint32
+}
+
+type TCPProcessor struct {
+	conns map[network.ConnectionTuple]connectionState
+}
+
+func NewTCPProcessor() *TCPProcessor {
+	return &TCPProcessor{
+		conns: map[network.ConnectionTuple]connectionState{},
+	}
+}
+
+func (t *TCPProcessor) updateSynFlag(conn *network.ConnectionStats, st *connectionState, pktType uint8, tcp *layers.TCP, payloadLen uint16) {
+	if tcp.RST {
+		return
+	}
+	// progress the synStates based off this packet
+	if pktType == unix.PACKET_OUTGOING {
+		st.localSynState.update(tcp.SYN, tcp.ACK)
+	} else {
+		st.remoteSynState.update(tcp.SYN, tcp.ACK)
+	}
+	// if any SynState has progressed, move to attempted
+	if st.tcpState == ConnStatClosed && (st.localSynState != SynStateNone || st.remoteSynState != SynStateNone) {
+		st.tcpState = ConnStatAttempted
+	}
+	// if both synStates are ack'd, move to established
+	if st.tcpState == ConnStatAttempted && st.localSynState == SynStateAcked && st.remoteSynState == SynStateAcked {
+		st.tcpState = ConnStatEstablished
+		conn.Monotonic.TCPEstablished++
+	}
+}
+
+// updateTcpStats is designed to mirror the stat tracking in the windows driver's handleFlowProtocolTcp
+// https://github.com/DataDog/datadog-windows-filter/blob/d7560d83eb627117521d631a4c05cd654a01987e/ddfilter/flow/flow_tcp.c#L91
+func (t *TCPProcessor) updateTcpStats(conn *network.ConnectionStats, st *connectionState, pktType uint8, tcp *layers.TCP, payloadLen uint16) {
+	payloadSeq := tcp.Seq + uint32(payloadLen)
+
+	if pktType == unix.PACKET_OUTGOING {
+		conn.Monotonic.SentPackets++
+		if !st.hasSentPacket || isSeqBefore(st.maxSeqSent, payloadSeq) {
+			st.hasSentPacket = true
+			conn.Monotonic.SentBytes += uint64(payloadLen)
+			st.maxSeqSent = payloadSeq
+		}
+
+		ackOutdated := !st.hasLocalAck || isSeqBefore(st.lastLocalAck, tcp.Ack)
+		if tcp.ACK && ackOutdated {
+			// wait until data comes in via SynStateAcked
+			if st.hasLocalAck && st.remoteSynState == SynStateAcked {
+				ackDiff := tcp.Ack - st.lastLocalAck
+				// if this is ack'ing a fin packet, there is an extra sequence number to cancel out
+				isFinAck := st.hasRemoteFin && tcp.Ack == st.remoteFinSeq+1
+				if isFinAck {
+					ackDiff--
+				}
+				conn.Monotonic.RecvBytes += uint64(ackDiff)
+			}
+
+			st.hasLocalAck = true
+			st.lastLocalAck = tcp.Ack
+		}
+	} else {
+		conn.Monotonic.RecvPackets++
+
+		ackOutdated := !st.hasRemoteAck || isSeqBefore(st.lastRemoteAck, tcp.Ack)
+		if tcp.ACK && ackOutdated {
+			st.hasRemoteAck = true
+			st.lastRemoteAck = tcp.Ack
+		}
+	}
+}
+
+func (t *TCPProcessor) updateFinFlag(conn *network.ConnectionStats, st *connectionState, pktType uint8, tcp *layers.TCP, payloadLen uint16) {
+	payloadSeq := tcp.Seq + uint32(payloadLen)
```

Comment:
Do we want to return early if the connection is already marked as closed?

Reply:
Oh that is a good idea

Reply:
Hm actually, I might need to keep running through this logic post-close to handle FIN retransmits in the future